lib/sales_engine.rb

```diff
@@ -153,145 +106,128 @@ def find_invoice_from_transaction(id)
-    end
+  def find_pending_customers_from_merchant(id)
+    transactions = find_transactions_from_merchant(id)
+    all_invoice_ids = transactions.flatten.map(&:invoice_id)
+    invoice_ids = find_successful_invoice_ids_from_transactions(transactions)
+    pending_invoice_ids = all_invoice_ids - invoice_ids
+    find_customers_from_invoice_ids(pending_invoice_ids)
   end
 
   def merchant_revenue(date="all")
-    merchant_revenue = []
-    merchant_repository.all.each do |merchant|
-      merchant_id = merchant.id
-      merchant_revenue << [find_revenue_from_merchant(merchant_id, date),merchant_id]
+    merchant_repository.all.map do |merchant|
+      [find_revenue_from_merchant(merchant.id, date), merchant.id]
     end
-    merchant_revenue
   end
 
   def merchant_items(date="all")
-    merchant_items = []
-    merchant_repository.all.each do |merchant|
-      merchant_id = merchant.id
-      merchant_items << [find_count_items_from_merchant(merchant_id, date),merchant_id]
+    merchant_repository.all.map do |merchant|
+      [find_count_items_from_merchant(merchant.id, date), merchant.id]
     end
-    merchant_items
   end
 
-  def find_most_revenue_from_merchant_repository(x)
-    merchant_ids = merchant_revenue.sort[-x..-1].collect {|i| i[1]}
-    merchants = merchant_ids.map do |merchant_id|
-      merchant_repository.find_by_merchant_id(merchant_id)
-    end
-    merchants.reverse
+  def find_most_revenue_from_merchant_repo(x)
+    merchant_ids = merchant_revenue.sort[-x..-1].map { |i| i[1] }.reverse
+    find_merchants_from(merchant_ids)
   end
 
-  def find_most_items_sold_from_merchant_repository(x)
-    merchant_ids = merchant_items.sort[-x..-1].collect {|i| i[1]}
-    merchants = merchant_ids.map do |merchant_id|
-      merchant_repository.find_by_merchant_id(merchant_id)
-    end
-    merchants.reverse
+  def find_most_items_sold_from_merchant_repo(x)
+    merchant_ids = merchant_items.sort[-x..-1].map { |i| i[1] }.reverse
+    find_merchants_from(merchant_ids)
   end
 
-  def find_revenue_by_date_from_merchant_repository(date)
-    merchant_revenue(date).collect {|i| i[0]}.reduce(:+)
+  def find_revenue_by_date_from_merchant_repo(date)
+    merchant_revenue(date).map { |i| i[0] }.reduce(:+)
   end
 
-  def find_most_revenue_items(x)
-
-    all_invoices = invoice_repository.all
-      invoice_ids = all_invoices.map(&:id)
-    all_transactions = invoice_ids.map do |invoice_id|
-      transaction_repository.find_all_by_invoice_id(invoice_id)
-    end
-
-    successful_ids = all_transactions.flatten.map do |transaction|
-      transaction.invoice_id unless transaction.result != ("success")
-    end
-
-    invoice_items = successful_ids.map do |invoice_id|
-      invoice_item_repository.find_all_by_invoice_id(invoice_id)
-    end
-
-    items_revenue = invoice_items.flatten.map do |id|
-      [id.quantity * id.unit_price,id.item_id ]
+  def find_successful_invoice_ids_from_transactions(transactions=transaction_repository.all)
```

Comment:
Line is too long. [92/80]<br>Surrounding space missing in default value assignment.